src/array/cpu/spmat_op_impl_coo.cc

```diff
@@ -190,16 +191,44 @@ std::vector<NDArray> COOGetDataAndIndices(
 
   std::vector<IdType> ret_rows, ret_cols;
   std::vector<IdType> ret_data;
-
-  for (int64_t i = 0, j = 0; i < rowlen && j < collen; i += row_stride, j += col_stride) {
-    const IdType row_id = row_data[i], col_id = col_data[j];
-    CHECK(row_id >= 0 && row_id < coo.num_rows) << "Invalid row index: " << row_id;
-    CHECK(col_id >= 0 && col_id < coo.num_cols) << "Invalid col index: " << col_id;
-    for (int64_t k = 0; k < coo.row->shape[0]; ++k) {
-      if (coo_row_data[k] == row_id && coo_col_data[k] == col_id) {
+  ret_rows.reserve(len);
+  ret_cols.reserve(len);
+  ret_data.reserve(len);
+
+  // NOTE(BarclayII): With a small number of lookups, linear scan is faster.
+  // The threshold 200 comes from benchmarking both algorithms on a P3.8x instance.
+  // I also tried sorting plus binary search.  The speed gain is only significant for
+  // medium-sized graphs and lookups, so I didn't include it.
+  if (len >= 200) {
+    // TODO(BarclayII) Ideally we would want to cache this object.  However I'm not sure
+    // what is the best way to do so since this object is valid for CPU only.
+    std::unordered_multimap<std::pair<IdType, IdType>, IdType, PairHash> pair_map;
+    pair_map.reserve(coo.row->shape[0]);
+    for (int64_t k = 0; k < coo.row->shape[0]; ++k)
+      pair_map.emplace(std::make_pair(coo_row_data[k], coo_col_data[k]), data ? data[k]: k);
+
+    for (int64_t i = 0, j = 0; i < rowlen && j < collen; i += row_stride, j += col_stride) {
```

Comment:
Will openmp help here?

Reply:
There are `push_back`'s in the loop body.

Reply:
Couldn't we first do a parallel sweep to find the locations and then use a parallel look up to fetch the actual data?

Reply:
For the current behavior we need to ensure that the returned edges are in the same order as given rows/cols.  I couldn't see benefit of parallel sweep when combining with `multimap`.  Correct me if I'm wrong though.

Reply:
The problem can be formulated as: given an array of keys A and an array of queries B, return two arrays BB and P such that A[P[i]] is the matched key of BB[i]. If A has no duplication, we want BB to be equal to B. Otherwise, just make sure that BB contains all the elements in B. The query array B can have duplicates. If both A and B have duplicate, the result is undefined. The size of B is usually much smaller than A.

We first check whether B has duplicates, which is efficient since the size of B is usually small.
* If B has duplicates, let each openmp thread work on a segment of array B. Each thread scans its own segment and writes the position of the matched key in A to P. You can build a hashmap of A (usually pretty costly) or just do a linear scan and stop at the first hit. No need to handle the case that A has duplicates. Throw error if any query misses.
* Otherwise, let each openmp thread work on a segment of array A. Each thread scans its own segment and writes the position to B of the matched query to a temporary array PB. Missed key has a special mark. Then remove the missed key in PB and get two arrays P and BB. Then sort both P and BB according to BB to get the final answer. You can optionally apply a linear scan on BB and throw error if any query misses. If we expect the size of BB to be similar with that of B, the sorting is efficient. Besides, building a hashmap on B would be a good choice.

Reply:
The problem is that we have two desired behaviors for `edge_id`:

Behavior 1: Return any edge that connects between `src` and `dst`, in the given order.
Behavior 2: Return all edges that connects between `src` and `dst`, potentially regardless of order.

I feel it's best implemented in two separate C++ interfaces, and I don't want the behavior selection to depend on "whether B has duplicates or not", because sometimes users may want to have behavior 2 even if B does not have duplicates (like in this example).  In fact, the behavior selection should depend on the argument "return_uv".  If "return_uv" is False, we assume the user wants behavior 1, and behavior 2 otherwise.

I would like to leave this refactor to a future PR (but of course before the next release) and for now I'm only focusing on making `edge_id` and `remove_edges` usable for the current example.  What do you think?

Reply:
My solutions are all for behavior 2. When will behavior 1 be desired?

> I would like to leave this refactor to a future PR (but of course before the next release) and for now I'm only focusing on making edge_id and remove_edges usable for the current example.

Could you add an item to our v0.5 roadmap?